### ISSUE TITLE
[UX] Fix for Manage blocks interface: cannot be used with keyboard navigation

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#4941]
**Description of the bug**

We received an email today from someone who is struggling with the manage blocks interface because they can not see. (The following is from a non-English speaker)

> Hi. I can not see. I am using my computer with NVDA reading the screen. Drag
and drop is done with the mouse. I have to use my computer only with the
keyboard. When using Drupal, there were line weights in block scrolling. I
was making the editing I wanted easy with the keyboard. Backdrop caught my
attention. However, I could not do anything with the mouse in the block
(settlement) area. Will the developer see this post? If it comes a line, we
will be very happy web developers who cannot see it. Thanks in advance.

The only obstacle to moving the site is the necessity of using the mouse.
The story is that everything can be done with the keyboard as if a mouse is
not attached.

**Steps To Reproduce**

To reproduce the behavior:
1. Go to the "manage blocks" page for Layouts
2. Attempt to reposition a block without using the mouse

**Actual behavior**

Unable to position blocks.

**Expected behavior**

Able to position blocks.

---

PR by @docwilmot and @tarekdj  : https://github.com/backdrop/backdrop/pull/3561 (drag n drop blocks, with cross-region drag. JS only)
PR by @docwilmot : https://github.com/backdrop/backdrop/pull/3556 (dropbutton link with dialog with to move blocks incl NOJS) Not currently preferred option,

**Comment by indigoxela** [comment]
I agree, that seems like a severe accessibility glitch.

It looks a bit tricky to me, though. The layout content form isn't a table, so we can't use tabledrag. We couldn't anyway, because the position in a layout should consist of two components:

- The region a block is in (by uuid?)
- The weight of the block within that region (there's no weight field yet)

That's a bit over my head. @docwilmot do you think, adding non-js positioning/sorting to the form is feasible?

**Comment by jenlampton** [comment]
@quicksketch recommended adding region and weight options to the block
configure forms. Not ideal, but it would at least make it possible to move
blocks.

**Comment by indigoxela** [comment]
> adding region and weight options to the block configure forms

The "Configure block" form actually does have a [region select list](https://github.com/backdrop/backdrop/blob/ac819913fdb7aadcec9c7579fd33920327526792/core/modules/layout/layout.admin.inc#L1291), currently only shown when js is disabled.
No weight field, though (and nothing yet, to handle a "weight" value in the layout form - as far as I can tell). Would such a setting conflict with the current drag-n-drop via jQuery UI sortable?

> Not ideal, but it would at least make it possible to move blocks.

I agree, that's not ideal, because it would require to edit several block settings to get the desired block order within a region. But better than nothing, for sure.

**Comment by docwilmot** [comment]
We could also look into hidden weight fields on the draggable blocks as well and use JS to update the hidden positioning field. Seems like that would be more helpful. Or probably both that, and a weight field on the configure forms. 

I'll see if I can do it.

**Comment by docwilmot** [comment]
Had a look yesterday, and this is doable, including weight fields on the drag-drop blocks. But realised something not previously mentioned: rearranging blocks is also impossible with JS disabled. 

To rearrange blocks without JS, we can do weight fields on the block configure forms as recommended, or we can have a separate page for re-ordering blocks. To see how that would work, create two layouts with the same path, then look at the layout list `admin/structure/layouts` and note that the `Menu settings` button changes to a dropbutton, with a `reorder` link.

Question tho: is a no JS solution that much necessary?

**Comment by jenlampton** [comment]
> rearranging blocks is also impossible with JS disabled.

This isn't a use-case we need to account for anymore since there is <1% of people on the internet that disable JS.

**Comment by docwilmot** [comment]
I only ask of course because we go to great lengths in all our UI to ensure a non-JS fallback. In fact, I believe the only thing that doesn't work well in Backdrop without JS is CKEditor.

**Comment by jenlampton** [comment]
> I only ask of course because we go to great lengths in all our UI to ensure a non-JS fallback.

Well, Drupal went to great lengths in all the UIs to ensure a non-JS fallback because in the past there used to be a much larger percentage of people/browsers who didn't always have JS. 

In Backdrop we would prefer to provide a non-JS fallback if it's easy, but we don't go to great lengths since it's not a real-world use case anymore.

**Comment by klonos** [comment]
Wouldn't it make sense to use a well-maintained 3rd-party solution, with accessibility already baked-in?

A little bit of quick research around the web resulted any many sites mentioning the term "Live regions".

> Live regions are HTML elements equipped with ARIA attributes that can be used to notify screen readers of content changes.

This was an interesting related read: https://medium.com/salesforce-ux/4-major-patterns-for-accessible-drag-and-drop-1d43f64ebf09 (subtitled: Building an interaction that works for both mouse and non-mouse users)

An interesting find, with a cool name 😄 https://github.com/schne324/dragon-drop (although only seems to handle lists - can each region in our blocks UI be converted to a list?) ...as well as a blog article showcasing it: https://www.smashingmagazine.com/2018/01/dragon-drop-accessible-list-reordering/

![image](https://user-images.githubusercontent.com/2423362/107988529-e5907200-6fd8-11eb-853f-87a41e9a2802.png)

Also: https://www.sitepoint.com/accessible-drag-drop ...with the code from that tutorial available here: https://github.com/sitepoint-editors/accessible-drag-and-drop

**Comment by indigoxela** [comment]
> In Backdrop we would prefer to provide a non-JS fallback if it's easy, but we don't go to great lengths since it's not a real-world use case anymore.

That absolutely makes sense to me. A little off-topic: it's easier to write functional tests with Simpletest, when there's a non-js fallback, but that's more a developer issue than a UX issue.

Web accessibility (using for instance aria-* attributes) is something, we should pay more attention. Personally, I find it hard to use these attributes in a reasonable and helpful way, though. Not because I don't care, but mostly because I don't know how to do it the right way. But it's something I really try to understand. I still hope, we can do it without using a third-party library.

**Comment by klonos** [comment]
> I still hope, we can do it without using a third-party library.

@indigoxela the 3rd party libraries in this case would be in order to get accessible keyboard navigation without re-inventing the wheel. As you said it yourself:

> ...because I don't know how to do it the right way.

It would simply take a considerable amount of time/research/effort to build an accessible, keyboard-navigable component ourselves from scratch 🤷 

Dependencies on 3rd party libraries in general are a two-edged blade: on one hand, projects with smaller communities get features "for free" (so they can focus their efforts on the actual features of the software they produce); on the other hand, it becomes a headache when these libraries become unmaintained or EoL (see jQuery 1.x, CKEditor 4.x etc.).

**Comment by indigoxela** [comment]
@klonos actually we're already using a third party library for drag-and-drop sorting - jQuery UI. To my understanding web accessibility in Backdrop is mostly accomplished with fallbacks like the weight (text) input items, as jQuery UI doesn't make use of attributes like role, draggable, aria-describedby and tabindex.

The fallback seems to work (in D and B) - if I read the cited feedback from the issue description correctly. Only the layout form (Manage blocks) currently offers no fallback and hence isn't accessible at all. - Which is what we're trying to fix here.

I'm glad that @docwilmot figured out a way to accomplish better accessibility on that form.

**Comment by docwilmot** [comment]
So, thinking, if we can use JS, a JS solution is probably better than what I was thinking, which was weight select fields on each block. Like "Move up" buttons on each block. Or really, a solution like what @klonos suggested above.

**Comment by indigoxela** [comment]
What would be the easiest way to achieve accessibility on that form?
Even if the solution wouldn't be perfect, it currently seems to me that anything could be an improvement.

@docwilmot sometimes the first thought is the best. As far as I know, we don't have any other form structured like "Manage blocks", so we probably can't re-use an existing design pattern.

**Comment by klonos** [comment]
> ...so we probably can't re-use an existing design pattern.

Yeah, I believe that this is the case.

**Comment by jlfranklin** [comment]
Similar to a hidden weight field on each block, a hidden "move to <region>" select would provide keyboard-only, programmatic, and no-JS support, in a UX style that is similar to the tabledrag hidden weight fields already in B and D.

I can drag a block from one region to another, reload the page without clicking save, and the old layout is restored, so without diving into the code, there seems to be an internal notion of "move this block to <region>" in the page already. All we'd be doing is exposing it as a JS-hidden form element.

**Comment by docwilmot** [comment]
This does not fully realise all the discussion above, but simply allows blocks to be moved *within their region* by keyboard actions. Blocks will need to be moved *between regions* using the block configure form. Its a start, and I'm thinking this could do till we find the time to add weight fields for regions and blocks, which might need more time than we've got.

**Comment by docwilmot** [comment]
FYI, tab to a block (in a region with more than one block ideally) then use up/down arrows to move it, or page up/down to move it to the top or bottom.

**Comment by indigoxela** [comment]
Nice! It works as described. :+1: 

I was able to move blocks around *within* a region only by using the keyboard. What I can not estimate is, if that's sufficient for people using a screen reader. For instance, is the focus of the currently active item clear enough?

One thing, that's potentially problematic is, that currently all blocks have the same "tabindex" of "0". Is this something that could get fixed?

**Comment by docwilmot** [comment]
> What I can not estimate is, if that's sufficient for people using a screen reader. For instance, is the focus of the currently active item clear enough?

Honestly don't know, would be great if someone with accessibility experience could advise what else we need there.
> One thing, that's potentially problematic is, that currently all blocks have the same "tabindex" of "0".

This is actually OK. `tabindex = 0` actually just enables tabbing to the block DIV, but the block stays within the normal document tab flow.

**Comment by jenlampton** [comment]
I have asked the original reporter to help us test:

```
Hello **********,

I'm happy that the NVDA screen reader program is working for you.

We have also been working on a fix for the layout page itself. Would
you be willing to help us test it?  If so, here is a "sandbox" website
where you can see the fix.

Website: http://3531.backdrop.backdrop.qa.backdropcms.org
Username: admin
Password: ngo3fndJ

Jen
```

**Comment by jenlampton** [comment]
Here's the reply we got:

> http://3531.backdrop.backdrop.qa.backdropcms.org/admin/structure/layouts/manage/dashboard
I did not notice any difference. There was no problem with Drupal. Got it
different on the backdrop. I move between regions with the add and remove
blocks method. The problem is that only blocks can be shifted up and down.
For example, user login and search blocks one above or another. I move
between footer-header by adding and removing.

Is there something specific I can ask this person to look for, to test?

**Comment by docwilmot** [comment]
I suspect a language issue here? Because I don't quite understand the response. 

In either case we should probably explicitly say what we've done: on the block add pages for standard layouts (haven't dealt with the flexible yet) you can now move blocks up and down using the up and down arrow keys or the page up page down keys. Moving blocks between regions is not yet possible by keyboard and requires using the block's configuration form. 

But we do not know what is needed to make this discoverable and accessible. Or if its helpful at all.

It may be helpful if we could reach one of the known D7 accessibility contributors.

**Comment by jenlampton** [comment]
> Moving blocks between regions is not yet possible by keyboard and requires using the block's configuration form

I tried the configure form, but could not see any way to move a block between regions. Is that a feature that needs to be enabled before it is visible? Or is it only visible to screen readers?

> It may be helpful if we could reach one of the known D7 accessibility contributors.

Perhaps we could ask @mgifford for his expertise?

**Comment by docwilmot** [comment]
> I tried the configure form, but could not see any way to move a block between regions. Is that a feature that needs to be enabled before it is visible? Or is it only visible to screen readers?

Drat you're right, I get confused between the hidden-for-accessibility vs non-JS. I believe the code only shows the region select form for non-JS browsing. Maybe we could just hide it with CSS instead.

**Comment by indigoxela** [comment]
Considering the reply:

> The problem is that only blocks can be shifted up and down.

That seemed to have worked, though! So, partly success.

And @docwilmot's comment:

> ...I believe the code only shows the region select form for non-JS browsing. Maybe we could just hide it with CSS instead.

Oh, that should have been there already! No, I didn't realize either, but thought it was something to implement later on.
Is that a job for css class "element-invisible"? Or what else could we use? Is there a way to cover both - no js and/or no mouse?

**Comment by mgifford** [comment]
Happy to help where I can @jenlampton. I was looking into this for Drupal a while back. Hadn't noticed the sitepoint reference, so added it to the list here https://www.drupal.org/project/drupal/issues/2920006

How closely is your layout tool connected to Drupal's at this point? There's a long thread here, with lots of children:  https://www.drupal.org/project/drupal/issues/3007978

This is a complicated issue. Partly the challenge is of basic navigation. Screen readers tend to have a linear path in a web page, so the dimension of left/right can be a bit of a challenge when most everything is essentially positioned as up/down. 

It's one thing to move the blocks, yet another to have that sense of the new relationship and the implications of that.

**Comment by indigoxela** [comment]
Another quick test in the sandbox: the form item in the block configure form (`$form['region']`) is now wrapped in "element-invisible", which should make it accessible to screen readers.

A screenshot with the form item made visible (turned off the class via browser dev tools):
![block-configure-dialog](https://user-images.githubusercontent.com/842066/108673192-8d2f0800-74e3-11eb-8245-0c500a7b4cd8.png)

A thought: if we added a toggle button there, we could make the form item accessible for devices without mouse, that don't provide access to hidden items otherwise.

@mgifford Welcome and many thanks for offering help, your know-how is highly appreciated.

Regarding code: there's a proof of concept [pull request](https://github.com/backdrop/backdrop/pull/3531), which also ships with a sandbox to conveniently test, what we currently have.

- Website: https://pr3531-ex1okk4srhtibgdr8v9gzpcqzq2s6nlk.tugboat.qa/
- Username: admin
- Password: e10fddb030f4

Just log in, go to (for instance) /admin/structure/layouts/manage/home and try to rearrange the blocks within the region, or try to change the region via an individual block configure form (dialog).

**Comment by mgifford** [comment]
Definitely some contrast issues on that page.  

Anyways, so much comes down to discoverability. With a mouse there is a clear indicator that you can move the elements up/down.  @indigoxela this isn't the case with a keyboard only user. 

The arrow keys didn't seem to work for me to re-organize the blocks.

**Comment by MustafaDogan78** [comment]
Hello there. I am the one who cannot see as said above. I can not speak English. I was not well understood because I did not speak English. I'm using google translate. It is not enough to just slide the blocks with the keyboard. It should be voiced with a screen reader. This is actually done with a js code. While the backdrop is set up, the installation of the modules is shown on the screen with informative texts. The screen reader reads them. Likewise, informative texts (eg balloons) are shown when the blocks are lost. As with Drupal, there may be hidden line weights. I don't expect it to go left and right. What is required is to go up and down in the same order. In the layouts, the place is changed with the add and remove method. I'm copying my screen reader link below. I made it portable for you. It is not installed on the computer. If you turn the screen reader on and try, you will understand the problem better. Keys to be used with a screen reader and their actions. (will be published together.)
insert space: Write mode is turned on. When moving blocks have to be opened. It closes when pressed again.
Insert q: Closes the program.
The program is open source. Thank you very much for your sensitivity. See you.
NVDA screen reader download link.
www.mustafadogan.net.tr/nvda.exe

**Comment by indigoxela** [comment]
@MustafaDogan78 welcome and many thanks for joining this discussion.

If I get it right, there should be an _active choice_ on the item to get moved, using the space key.

The workflow could be as following:

1. navigate to the block item via tab key (no change regarding that)
2. actively select the item with the space tab (selecting an item also provides some informational text, marks the item as "movable")
3. move the selected block item within its region with the arrow up/down keys
4. actively deselect the block item by hitting the space tab again (the item is not marked as "movable" anymore)
5. the rest stays unchanged

Did I summarize it correctly, or did I miss anything important?

@docwilmot does this seem feasible with your current approach?

**Comment by MustafaDogan78** [comment]
No problem for it done with the keyboard. The problem is that this is not known except by eye. The screen reader doesn't read what is there. Wherever the block is headed, it is sufficient if an information about it appears on the screen for a short time and disappears. For example, maybe a one-second message, such as the user input has been scrolled to the top of the search block, or the welcome block has been scrolled under its main.

**Comment by MustafaDogan78** [comment]
I came, the conversation stopped. Just solved the problem with the keyboard. The accessibility issue remains. If I can't understand Google translate, I'll give money to a Turk who speaks English and make a tool. But it's not nice that the problem continues with this screen reader. In Turkey, nine out of ten people use Drupal. I showed them Drupal. Association sites, government agency sites, libraries; it was always done with Drupal. Now there are problems with updating. I'll direct them to the backdrop, but they can't use it that way. There was hope here. I had hope that the problem would be solved. But the subject seems to have cooled. I've been translating at night until morning for ten days. Along with the background kernel, I also Turkified the following modules. I'm translating it with Translate and showing it to my school friend who's doing a PhD in America. Then I add it to the interface. It's good to be a part of something. Beautiful enough to translate without knowing the language. But if this problem is not solved, the Backdrop will not work for those of us who do not see it. My translations will be our gift if anyone uses them.
audiofield, backup_migrate, basic_cart, captcha, comment_notify, eu_cookie_compliance, flexslider, forum, riddler, smtp, telephone, trigger_actions, user_prune, webform, xmlsitemap.
I also translated the following themes into Turkish.
lateral, basis.

**Comment by jenlampton** [comment]
Wow, that's fantastic! Thank you for all the work you have done translating @MustafaDogan78. Please rest assured that we are still working on this issue, but since everyone is volunteering their own time, the speed at which issues progress will increase and decreases depending on what we all have going on in our lives. We have not given up!

**Comment by MustafaDogan78** [comment]
I thought you had left now. The blocks moved with the keyboard, I thought you said it was okay. I'm glad you're giving up.

**Comment by MustafaDogan78** [comment]
I thought of something.
I do not know if it can be done.
On the block configuration page, get a question like this.
Where should this block appear?
below is the selection box.
one more question after the election.
Where should this block be in this area?
or under which block should this block be in this area.
or on
like first state then city selection
When you press the save button, the process is done automatically.
So let it end with the block configuration instead of the internal layout.
Could such a thing be possible?


25 Şubat 2021 10:54:35 PM Jen Lampton <notifications@github.com> yazdı:

**Comment by docwilmot** [comment]
I just had a better thought: how about a "Move" link in the block contextual links? Opens a dialog with noJS fallback, with a region select list that loads a tabledrag list of the blocks in that region! Keyboard accessible, noJS sorted. Will work on a PR ASAP.

Advocating...

**Comment by MustafaDogan78** [comment]
Seems logical. However, the scenario is different, the result may be different. It is important that that list can be read with a screen reader. Now the situation is not bad actually. the only problem is the screen reader doesn't read the action. Just like the Backdrop setup, as the Modules are installed, it writes a message telling you which module was installed. Screen readers read it easily. When moving blocks, the text to be extracted with a js code that specifies which block the block goes up or down also solves the problem. The important thing is to be resolved. I'm just saying advice. Whichever method will be easy, that method should be used. The important thing is to get the job done.

**Comment by docwilmot** [comment]
This latest PR allows noJS and keyboard manipulation of blocks. BUT sadly, after I've done this, tested with screenreader and discovered D7/Backdrop tabledrag isn't accessible either. But I think it's an improvement.

I fee we could either add some more accessibility features to this or make an issue separately to add those features generally.

**Comment by MustafaDogan78** [comment]
NVDA screen reader is open source. It has plugins too. There is also a drag and drop plugin. Even that extension gets the job done, but not everyone has to use NVDA. Not everyone has to use the plugin either. We need to find the solution within ourselves, in the Backdrop. The link to the plugin is here.
https://github.com/javidomerox/DragAndDrop

**Comment by indigoxela** [comment]
@docwilmot the no-js fallback still has some glitches, but I see the direction it's going and I like it.

For the sighted it's currently duplicating the drag-and-drop functionality, which might not be ideal, but probably doesn't hurt (?). Maybe we find a smart way to hide the link for devices that don't need it?

Regarding tabledrag and accessibility (make reordering by keyboard easier out of the box): is it supposed to be a follow-up issue or should it get handled here? Backdrop uses tabledrag all over the place, so making it more accessible will fix many problems at once.

**Comment by MustafaDogan78** [comment]
Of course, we can open a new topic for general accessibility issues.
For example, there is a problem with permissions.
how to get roles first, then checkboxes.
However, it should say the roles along with the checkboxes.
Of course, a new topic can be opened in which all of these will be discussed.


6 Mart 2021 14:47:57 indigoxela <notifications@github.com> yazdı:

**Comment by MustafaDogan78** [comment]
Also of course it is nice to have something to do with the keyboard.
With a javascript code, if it shows messages on the screen indicating where 
the blocks are going, even if only for 1 second, screen readers will 
announce it.
for example the user block has been moved under the main menu ...


6 Mart 2021 14:47:57 indigoxela <notifications@github.com> yazdı:

**Comment by indigoxela** [comment]
For the records: Drupal 9 uses Drupal.announce for that, which in turn uses aria-live and aria-busy attributes. Here's a link to [their source code](https://git.drupalcode.org/project/drupal/-/blob/9.2.x/core/misc/announce.js).

Using something like that could provide feedback for screen readers about what happened or changed.

**Comment by MustafaDogan78** [comment]
If I remember correctly, drupal 9 had no problem.
I will still install it on the wamp server and check it out.
None of the drupan versions have accessibility problems.
I did not see young people in Turkey for years and directs sites they did.
Although there is little support for Turkish
used more than wordpress.
now they are thinking how to upgrade because of me.
Unfortunately, the promises of upgrades in versions 8 and 9 could not be 
fulfilled.
I can no longer tell them anything about it.
because I lost hope too.
her defasında yeni baştan site yapmak yorucu oluyor.
It is tiring to build a new site every time.
I recommend backdrop cms, but these issues should be resolved.
They will not want to switch to a system that they cannot use.


6 Mart 2021 15:53:08 indigoxela <notifications@github.com> yazdı:

**Comment by docwilmot** [comment]
I've updated the original post with the three "solutions" we have so far:
(note these are my thoughts, and I suspect a person with a11y experience may disagree on priorities and preferences)
- my original PR adds drag n drop to the blocks. Needs better discovery for accessibility
- a PR by @tarekdj building on the above, that adds moving blocks *between* regions. Also needs discoverability.
- another PR by myself, which adds a dropbutton links and a dialog to rearrange bocks or move them between regions. I think this option is closest to the D8/9 layout builder.

All of these would benefit from a core API for a11y announcements like Drupal's `Drupal.announce()`.
I'm not sure how best to proceed. But I really think we should prioritise this as Laout blocks cannot be rearranged at all right now and is unusable by keyboard users, as @MustafaDogan78 says above.

**Comment by stpaultim** [comment]
If you are looking for more involvement on this issue, I think that a screenshare during an upcoming dev or UX meeting might be helpful. I think a live demo and the ability to ask questions would help me figure out how I can best help or provide feedback. Is this something that could be demonstrated during a meeting?

**Comment by MustafaDogan78** [comment]
Moving blocks between layouts is not difficult.
We can handle it by removing blocks from one order and adding blocks to 
another order.
What matters is that the blocks in an order go up or down.
it was not possible with the keyboard now this has fixed.
All that remains is to get screen readers to voice it.


6 Mart 2021 22:04:26 docwilmot <notifications@github.com> yazdı:

**Comment by MustafaDogan78** [comment]
Can't hidden row weighting like drupal be done?


6 Mart 2021 22:04:26 docwilmot <notifications@github.com> yazdı:

**Comment by MustafaDogan78** [comment]
Now I saw it on the test site. The positioning of the blocks between regions has been solved. Setting up from the block configuration window. The progress is very good. I'm really glad. Thanks to those who worked on it. There is 1 job left to do. a text message on the screen to tell you when blocks are dragging up or down. (If there are line weights hidden, -1 1 text messages are not needed.)

**Comment by MustafaDogan78** [comment]
I thought. How are the move down and move up links below the list additional actions button?

**Comment by docwilmot** [comment]
@mgifford @MustafaDogan78 could you please advise which of the options [in this comment](https://github.com/backdrop/backdrop-issues/issues/4941#issuecomment-792026585) would be most helpful? You can test in the sandboxes linked [in the first post](https://github.com/backdrop/backdrop-issues/issues/4941#issue-804832339). I'm not actually sure which one to work on and which to stop.
> I think that a screenshare during an upcoming dev or UX meeting might be helpful. 

Hopefully once I get some direction, we can do that.

**Comment by mgifford** [comment]
Totally agree with the Drupal's Drupal.announce() suggestion.  Other than that I'm not sure. Where is the latest sandbox?

**Comment by MustafaDogan78** [comment]
Which can be done easily? I just wrote the solutions I thought about. There's already a Hide button inside the blocks. It would be enough if the line weights were placed under it.

**Comment by MustafaDogan78** [comment]
Actually, just this is enough. I noticed it while wandering at this address.
http://3531.backdrop.backdrop.qa.backdropcms.org/admin/structure/layouts/manage/home
Here's something like this.
"List additional actions"
If a feature like the one below is added under this button.
http://3531.backdrop.backdrop.qa.backdropcms.org/admin/config/content/formats
There is also something like this there.
Weight for Filtered HTML
0
Configure
Weight for Raw HTML
one
Configure
Weight for Plain text
10
Configure
This is a complete solution.

**Comment by docwilmot** [comment]
@MustafaDogan78 it seems you mostly want the ability to move blocks up and down, am I right? The suggestion you give does not allow moving the block between regions.

**Comment by stpaultim** [comment]
> Where is the latest sandbox?

The three PRs are listed in the [original post](https://github.com/backdrop/backdrop-issues/issues/4941#issue-804832339) (at the end of the original post) in this issue. I assume the highest PR# is the latest.

**Comment by MustafaDogan78** [comment]
Moving blocks between regions has already been resolved. Even if it didn't, I'd move it between regions by doing add and remove. Yeah, the important thing is that it slides up and down.

**Comment by indigoxela** [comment]
> So my current preference is https://github.com/backdrop/backdrop/pull/3561

Same here! So cool, well, almost. :wink: 

Blocks can get moved *within* a region by using up and down keys, and moved between regions by "k" and "j" keys. No extra form - everything happens on the main "Manage blocks" page.

But - the blocks get duplicated (copied) when supposed to get moved, both when trying to move within a region and when trying to move to a different region. I could reproduce that several times in the sandbox, ending up with several instances of "Custom block 1" in the home page layout.

@jenlampton you chose milestone 1.19.0 - shouldn't that be 1.18.2 for a bugfix?

**Comment by jenlampton** [comment]
This is going to need automated tests, I suspect, so will likely need to go
into 1.19.

**Comment by indigoxela** [comment]
> This is going to need automated tests, I suspect, so will likely need to go

Oh, I didn't know that automated tests require a minor version milestone... (?).

**Comment by docwilmot** [comment]
> But - the blocks get duplicated (copied) 

Thanks, I tested this a million times but never saved it so never saw that. I think I've fixed that now.
> This is going to need automated tests, I suspect, so will likely need to go

I don't know how to do automated tests for a purely JS solution for one, and second, this is a bugfix, the UI is unusable with the keyboard as is now. I thnk we need to do this right away.

**Comment by jenlampton** [comment]
> Oh, I didn't know that automated tests require a minor version milestone... (?).

That seems to be the current pattern. I'm not sure if it's a hard rule.

> I thnk we need to do this right away.

Well, let's give it a shot and see what the committers think!

**Comment by MustafaDogan78** [comment]
And I say he should leave now. The update happens for 2 reasons.
1. bug fixes.
2. Security issues.
There are fixes a bug here. 1.19 would have been expected if the feature had been added.

**Comment by quicksketch** [comment]
Based on the code complexity, I don't see why this needs to wait for 1.19.0. As a UX issue it's an improvement. As an accessibility issue, I would consider it a bug. Both could be used as justification to put into a bugfix release.

**Comment by indigoxela** [comment]
@docwilmot many thanks for updating the pull request, the block duplication problem is fixed now. Works for me. Note: there's no feedback for screen readers yet, about what happened. That will be implemented, as soon as the new "backdrop.announce" API is available.

Should we add some info about which keys do what? How will people know that "k" and "j" move blocks to next/previous region? Are there best practices for that?

**Comment by MustafaDogan78** [comment]
I can't speak English. I had to write here. I'm sorry. Using translation is very difficult. I have to write this.
I'm gonna write a script now. It's not about accessibility. It's a general problem. Let's create a special block. The configuration of this block is very limited. It cannot be configured without going to the layouts page and adding it to a region. For example, if the block is only shown on certain pages. For example, the contract will be added to the user registration form. I can't do this directly in the custom block insertion area.

**Comment by MustafaDogan78** [comment]
I'm one of those people who says it should be added right away. It should be updated as version 1.18.2. By May, it would be harmful if even one person tried it and gave up because I couldn't use it. We are not yet a cms with millions of users. This is a bug fix and does not expect errors.

**Comment by alanmels** [comment]
Alan Mels: Daha önce fark etmediğim için özür dilerim. Hadi Atamızın diline geçelim, anlat kardeş ne yapmaya çalışıyorsun, bakalım yardım edebilir miyim.

Alan Mels: İstersen zoom falan yapabiliriz, bazen yazışmaya konuşma tercihli ya.

**Comment by MustafaDogan78** [comment]
Dört tane gönderi yazdım. Hepsini tek tek yazayım mı? Yoksa son yazdığımı mı soruyorsun? Aşağıya şimdiye kadar yazdıklarımı tek tek sıralıyorum.
    1. Ben körüm. bilgisayarımı ekran okuyan bir programla kullanıyorum. bloklar sürükle bırak ile taşınıyor ve ben sadece klavye kullanmak zorundayım. Çeviriyle de olsa anlatmayı başardım ve klavye ile taşıma işini çözdüler. Hepsine teşekkürler. Ancak şimdi de ekran okuyucu ile okumuyordu. son zamanda onu da çözdüler galiba.
    2. Bir problem de izinler kısmında. Drupal kullanırken, onay kutusuyla birlikte rol okunurdu. Şimdi ekran okuyucu önce rollleri sırayla okuyor. Sonra da onay kutucuklarını işaretli, işaretli diye okuyor.
    3. On gece boyunca sabahlara kadar türkçe çeviri yaptım. Daha önce drupal kullandığımdan, hem google çeviri, hem yandex çeviri ile çevirileri yaptırıyorum. sonra Drupal’dan kalan alışkanlıkla mantık yürütüyorum. Sonra da ingilizce bilen bir arkadaşıma gözden geçirtiyorum. Bu şekilde 2400 dizi çevirisi önerdim. Şu an kullandığım backdrop sitemde (www.mustafadogan.net.tr) türkçe çeviri oranı %94. Halbuki https://localize.backdropcms.org/translate burada %48. Ne zaman eklenir bilemiyorum. Binlerce dizi. zaman alması normal.
    4. Son yazdığımsa, özel bloklarda blok oluşturuyoruz. Yapılandırma seçenekleri çok sınırlı. Mesela bir ziyaretçi siteye üye olacağı zaman, ona bir sözleşme göstereceğiz. yeni blok oluşturduk. Blok başlığı üyelik sözleşmesi. Blok gövdesi, lütfen sözleşmeyi okuyunuz... Vs... Normalde blok altında bu bloğun nerede görüneceği ayarlanır. belli bir sayfada görünsün deriz ve oraya user/register yazarız. Ama backdrop cms özel blokta bu yok. İllede layouts kızmına gidip bloğu bir bölgeye ekleyip sonrasında yapılandır linkinden bunu yapmak zorundayız. Halbuki özel blok oluşturulduğu an bu yapılabilmeli.
Bu cevaba çok sevindim. ilgine çok teşekkürler. Seninle tanışmak isterim. Türkiye’den selamlar.

**Comment by alanmels** [comment]
Mustafa kardeş,

Şimdilik yazdıkların ile ilgili kısa bir şekilde cevaplamaya çalışıyım, fakat bence WhatsApp yada Zoom kullanarak bir konuşsak daha bir efektif şekilde yardım edebilirdim.

1. Anladığima göre birini problem çözülmüş, değil ise nasıl yardımcı olabilirim?
2. Backdrop Drupal'dan türediğine rağmen bir çok şeylerde farklıdır. Fakat izinler baya benzer olmalı. İşaretli derken yapamadığın tam nedir? 
3. Tercümeler için çok teşekkürler, aslında ben seni Türkçe için admin olarak onaylamıştım, fakat kendin direkman yeni tercümeleri yukleyebilicek misin emin değilim. Tercümeler idare edilen site da baya yeni, bazı fonksyonlar tam olarak çalışmıyor. Fakat konuşalım bir audo olarak ve Türkçe tercümeleri istediğin veya müsait olduğun zaman yukleyebilmen için mutlaka yolun boluruz.
4. İşte bloklar Backdrop'da Layout'lar geldiği için biraz farklı idare edilirler. Fakat merak etme: senin bulamadığın görünme ayrları ordadır. Mutlaka yardım edeçeğim, ama audıo görüşmesi yapalım, senin için daha kolay olur.

**Comment by MustafaDogan78** [comment]
Hızlı geri dönüş için teşekkür ederim. Ben WhatsApp kullanmıyorum. Ayrımcılık yaptıkları için hesabımı silip telefonumdan kaldırdım. Ancak telegram ve signal kullanıyorum. Telefon numaramı aşağıya yazıyorum. Sesli görüşürsek ben de memnun olurum.
Tel: +90(542)674 5233

**Comment by alanmels** [comment]
We had a nice conversation with Mustafa and I've got the main problem he's facing. I'll try to create PR later today.

**Comment by oadaeh** [comment]
> > When using Drupal, there were line weights in block scrolling. I
> > was making the editing I wanted easy with the keyboard.
> > ...However, I could not do anything with the mouse in the block
> > (settlement) area.
> > ...If it comes a line, we
> > will be very happy web developers who cannot see it.

I'm sorry I'm coming to this late.

While I see that everyone is trying to make it so that the blocks are movable with a keyboard, I think another problem is being mentioned in the above quoted message.

The issue that I think is most closely related is #1632.

**Comment by alanmels** [comment]
As a visually impaired person Mustafa is experiencing multiple obstacles in using Backdrop, but instead of adding on top of this already long page, I decide to tackle the problems he explained one by one in separate issues and respective PRs. Please review this one https://github.com/backdrop/backdrop-issues/issues/4996

**Comment by MustafaDogan78** [comment]
Alan kardeş sana zahmet bunu ingilizceye çevirir misin? 
Jason Flatt Doğru anlamış. Asıl problem Jason Flatt  arkadaşımızın anladığı. Ben gördüğüm bütün eksikleri yazdım. Bu da yanlış anlamaya sebep olmuş. Jason Flatt arkadaşımızın söylediği sorun asıl sorunumuz. Benim bundan sonra erişilebilirlik dışındaki sorunları yazmamam gerekecek galiba. Jason Flatt arkadaşın söylediği problemi çözüldü zannettiğimizden, signal sobetimizde  üzerinde durmadık.

**Comment by alanmels** [comment]
Mustafa, bende gece saat 3 oldu, seni yarin arayacgim. Simdilik cabuk bir ceviri.

Guys, Google Translate sometimes is not as much reliable, especially when it comes to agglutinative languages like Turkish. Mustafa is asking me to translate.

He thinks that Jason Flatt got it just right apparently meaning @oadaeh's post on https://github.com/backdrop/backdrop-issues/issues/4941#issuecomment-798830167. He believes because he tried to write up all different issues at once and that lead to misunderstandings. Mustafa thinks Jason Flatt's finding is exactly what his main hurdle is. And he keeps explaining this time to myself that while we discussed https://github.com/backdrop/backdrop-issues/issues/4996 on Signal today he missed to explain to me his main problem. 

I'm gonna talk to Mustafa tomorrow to come up with more orderly reporting of his issues, because as I see lot's of different issues are poling on top of each other. One thing should be addressed at a time. So please, guys, review https://github.com/backdrop/backdrop-issues/issues/4996 as that one improves usability a bit, which is especially important for visually impaired users.

**Comment by indigoxela** [comment]
One general recommendation: this issue thread is growing quickly and handles lots of problems, that aren't actually the topic here.

Can we please try to focus better on actual topics? If a new issue is needed, don't hesitate to create one, but try as much as possible to keep issues focused on only one problem.

Reason for this: It's easier for developers to solve clearly defined problems, than ambiguous ones. The same applies to testers and reviewers.

I'm aware of the language problem here, but it doesn't make things easier for us, if - possibly important - parts of the discussion happen in a language, hardly anyone of the people involved speak.

**Comment by MustafaDogan78** [comment]
indigoxela I caused the confusion. I'm sorry. I'll be more careful later. Alan already speaks Turkish. He helps with language and other matters. Now dialics will be healthier.

**Comment by alanmels** [comment]
We've had another call today with Mustafa and agreed on two things:

1. I explained him that it could take some time, but the Backdrop community will do it's best to mitigate the issues he raised, and asked him to avoid mixing issues and to stick with just one issue at a time.
2. His answers sent from his e-mail client program seem to be adding additional lines, adding confusion to already badly translated posts, so finding out it wouldn't be any more difficult for him to actually login to and follow-up with issues directly on GitHub, I asked him to do exactly so.

Hopefully, now we can help him in more structured ways. And since I noticed Mustafa's threads too late (due to lack of time I usually follow up issues that only myself created) and since this thread is really complicated I have couple questions if someone could please enlighten me.

1. Could someone review https://github.com/backdrop/backdrop-issues/issues/4996 as in Mystafa's believe that slight change would make things easier. The wording I came up with might be not ideal, so feel free to suggest changes on PR, but let's get it done.

2. We made some tests with Mustafa today and were not able to move blocks without mouse on the last dev branch, so as I understand changes discussed above have not been committed. Could you, guys, confirm I should follow https://github.com/backdrop/backdrop/pull/3561? The tagboat says "This preview is suspended." on that PR, so I will just manually apply changes for now to run tests, but is that the right one?

**Comment by docwilmot** [comment]
> Were not able to move blocks without mouse on the last dev branch
I should follow backdrop/backdrop#3561? 
The tagboat says "This preview is suspended." 

THis is the correct PR, and no that PR is not yet committed to core, but it is likely to be, hopefully soon. The Tugboat sandboxes often say "suspended" for a few seconds, but then they wake up spontaneously. I just tried the latest sandbox and was able to log in, and was able to move blocks with keyboard. FYI the keys are:

- Up/down arrows to move blocks within regions
- Pg Up/Pg Dn buttons to move blocks to the top or bottom of a region. Keys T and B do the same
- J and K to move blocks to the next/previous region

Of course Tab to select the block before it will move. Added a black shadow highlight on selection to improve visibility, but that may not be the final CSS. 

Finally, we're hoping that even before committing this, we will be able to add our new Backdrop announce feature to allow more verbose ARIA descriptions (to announce which block is selected, its position in the queue, where it was moved to etc etc. Will be happy to get comments from @MustafaDogan78 via @alanmels to guide us.

I'm very pleased that we're working on this, to be honest. Just waiting on a core reviewer/committer. Hopefully soon.

**Comment by alanmels** [comment]
@docwilmot, thank you very much for the detailed summary. I will try to test the branch later today. I am currently away from my desk with MBP, so using Ubuntu built on old MBP, keys are different, but I'll google it up to find equivalents.

> Finally, we're hoping that even before committing this, we will be able to add our new Backdrop announce feature to allow more verbose ARIA descriptions (to announce which block is selected, its position in the queue, where it was moved to etc etc. 

Would be wonderful.

> I'm very pleased that we're working on this, to be honest. Just waiting on a core reviewer/committer. Hopefully soon.

Same. I deeply felt during our talks with Mustafa how challenging (seemingly simple to other users) things visually-impaired persons. So let's make Backdrop no less functional than Drupal 7 is in this respect.

**Comment by alanmels** [comment]
Luckily tagboat was there for me this time and I've quickly tested to confirm: moving blocks works perfectly well! Great job, guys! Looking forward for it to be committed.

**Comment by alanmels** [comment]
Mustafa have just confirmed he has downloaded https://github.com/docwilmot/backdrop/archive/keyboard-block-move.zip and built the test website locally and successfully moved blocks around. He is happy and appreciative to you, guys, for the great job done. I've explained him if he waits some time then we will also enjoy ARIA descriptions.

**Comment by MustafaDogan78** [comment]
Keyboard shortcut keys for block movements my advice.
="d" for down,
="u" for up,
="l" for left
="r" for right.

**Comment by indigoxela** [comment]
@MustafaDogan78 Many thanks for your suggestion. Another question: what would be optimal keys to move a block to the previous or next region? Do "p" and "n" seem reasonable, or does this conflict with anything you use in NVDA?

**Comment by MustafaDogan78** [comment]
The insert key is used together with NVDA for shortcuts. No problem with NVDA if there is no insert button.

**Comment by quicksketch** [comment]
I think using letters for movement is going to be hard to discover. We can't reliably depend on a particular language (such as "R" for "Right" or "N" for "Next") or a keyboard layout (in the U.S. the WASD keys are common). I think using the arrow keys would be easiest for discoverability, have the fewest internationalization issues, and match our keyboard support for draggable tables (such as when organizing a menu).

Up/Down - Move up and down
Left/Right - Move between previous and next regions

I don't have an opinion on Page Up/Down having special behavior.

**Comment by MustafaDogan78** [comment]
I wrote it when I saw the" j "and" k " keys. of course, the direction keys make more sense.

**Comment by quicksketch** [comment]
Rather than wait on https://github.com/backdrop/backdrop-issues/issues/4985, let's get this in and do a follow-up. This issue is really long already. @docwilmot already has made all the suggested changes in the pull request and it looks ready.

I have merged https://github.com/backdrop/backdrop/pull/3561 into 1.x and 1.18.x. We'll be making the 1.18.2 release later today.

Thanks so much to everyone involved in this issue! The outcome is really excellent. A big thank you to @MustafaDogan78 for your patience, and @alanmels for helping translate. @docwilmot  and @tarekdj: thanks for your implementation work. And @indigoxela, @jenlampton, @klonos, and @mgifford for feedback and testing.

**Comment by MustafaDogan78** [comment]
The news is gratifying. I did not test it. I would love to test it. I hope it happened. Thanks to everyone who made the effort.

**Comment by quicksketch** [comment]
The ARIA announcement functionality is also now in Backdrop, but we have not integrated it into the block user interface. We will do that work in the follow-up issue https://github.com/backdrop/backdrop-issues/issues/5021.

**Comment by MustafaDogan78** [comment]
I tried and unfortunately negative. I'm adding a test video below. The screen reader still doesn't tell where the blocks go.
https://www.mustafadogan.net.tr/layout.mp4

**Comment by MustafaDogan78** [comment]
It is necessary to wait a little longer. I'll make a better test video tomorrow. I think there is a problem. There are also differences with regard to internet browsers.

**Comment by quicksketch** [comment]
Thank you for the video, it was very helpful.

> The screen reader still doesn't tell where the blocks go.

The is correct. My apologies. The keyboard navigation works (your video confirms this). And we built the system for announcements in https://github.com/backdrop/backdrop-issues/issues/4985.

However, we have not yet incorporated the two systems together yet. So while the blocks move with the keyboard, we are not yet announcing the new positions. We will continue working on this in the new issue filed at https://github.com/backdrop/backdrop-issues/issues/5021

**Comment by MustafaDogan78** [comment]
The test was done with 1.18.2 on wamp. If there is another package that I can test, you can send a download link to iletisim@mustafadogan.net.tr. thanks.

**Comment by alanmels** [comment]
Guys, just so you know - while talking to Mustafa last time, I've found out he was confused how to download a particular feature branch, so when I provided him the direct link to zip file, he was immediately able to download it and run his tests. So for any new PR please provide him the direct links such as https://github.com/docwilmot/backdrop/archive/refs/heads/announce.zip.
